test/Test/BaseDAO/Token.hs

```diff
@@ -0,0 +1,106 @@
+    , nettestScenario "can mint tokens from any accounts" adminMintScenario
+    ]
+  , testGroup "Other:"
+    [ nettestScenario "can call transfer tokens entrypoint" transferContractTokensScenario
+    , nettestScenario "can call token address entrypoint" tokenAddressScenario
+    ]
+  ]
+
+adminBurnScenario :: (Monad m) => NettestImpl m -> m ()
+adminBurnScenario = uncapsNettest $ do
+  ((owner1, _), _, dao, admin)
+    <- originateBaseDaoWithBalance DAO.defaultConfig
+          (\o1 _ ->
+              [ ((o1, 0), 10)
+              , ((o1, 1), 10)
+              ]
+          )
+
+  callFrom (AddressResolved admin) dao (Call @"Burn") (DAO.BurnParam owner1 DAO.unfrozenTokenId 100)
+    & expectCustomError #fA2_INSUFFICIENT_BALANCE (#required .! 100, #present .! 10)
+  callFrom (AddressResolved admin) dao (Call @"Burn") (DAO.BurnParam owner1 DAO.frozenTokenId 100)
+    & expectCustomError #fA2_INSUFFICIENT_BALANCE (#required .! 100, #present .! 10)
+
+  callFrom (AddressResolved admin) dao (Call @"Burn") (DAO.BurnParam owner1 DAO.unfrozenTokenId 10)
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner1 0
+  callFrom (AddressResolved admin) dao (Call @"Burn") (DAO.BurnParam owner1 DAO.frozenTokenId 5)
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 5
+
+adminMintScenario :: (Monad m) => NettestImpl m -> m ()
+adminMintScenario = uncapsNettest $ do
+  ((owner1, _), _, dao, admin)
+    <- originateBaseDaoWithBalance DAO.defaultConfig
+          (\o1 _ ->
+              [ ((o1, 0), 0)
+              , ((o1, 1), 0)
+              ]
+          )
+
+  callFrom (AddressResolved admin) dao (Call @"Mint") (DAO.MintParam owner1 DAO.unfrozenTokenId 100)
+  checkTokenBalance (DAO.unfrozenTokenId) dao owner1 100
+  callFrom (AddressResolved admin) dao (Call @"Mint") (DAO.MintParam owner1 DAO.frozenTokenId 50)
+  checkTokenBalance (DAO.frozenTokenId) dao owner1 50
+
+transferContractTokensScenario :: (Monad m) => NettestImpl m -> m ()
+transferContractTokensScenario = uncapsNettest $ do
+  ((owner1, _), (owner2, _), dao, _) <- originateBaseDao
+  let addParams = FA2.OperatorParam
+        { opOwner = owner1
+        , opOperator = toAddress dao
+        , opTokenId = DAO.unfrozenTokenId
+        }
+  callFrom (AddressResolved owner1) dao (Call @"Update_operators") [FA2.AddOperator addParams]
+
+  let transferParams = [ FA2.TransferItem
+            { tiFrom = owner1
+            , tiTxs = [ FA2.TransferDestination
+                { tdTo = owner2
+                , tdTokenId = DAO.unfrozenTokenId
+                , tdAmount = 10
+                } ]
+            } ]
+      param = DAO.TransferContractTokensParam
+        { DAO.tcContractAddress = toAddress dao
+        , DAO.tcParams = transferParams
+        }
+
+  -- TODO: Figure out why FA2_NOT_OPERATOR occurs
+  callFrom (AddressResolved owner1) dao (Call @"Transfer_contract_tokens") param
+   & expectCustomError_ #fA2_NOT_OPERATOR
```

Comment:
I've dropped the prefix for various params to match the spec.

I also rebased and use `nettestScenario` instead.

I've used `contractCalling` to implement `transferContractTokens` entrypoints, but there are 2 problems that I found:
1. `contractCalling` returns `Maybe a`, in case of error, which error should I use in this case.
2. From my understanding, the above test should work fine. However this causes `FA2_NOT_OPERATOR`. 

So my question is, when we use `contractCalling` from a `contractA` to call `Transfer` entrypoint of `contractB`. Is the sender address the contractA's address itself or the original sender of contractA ?

Reply:
> contractCalling returns Maybe a, in case of error, which error should I use in this case.

I think there are 2 options:
1. Add a new error and fail with it.
2. Take `contract [TransferItem]` parameter instead of `address`. If you have a `contract`, you don't need to call `contractCalling`.

> So my question is, when we use contractCalling from a contractA to call Transfer entrypoint of contractB. Is the sender address the contractA's address itself or the original sender of contractA ?

It should be `contractA`'s address, so I am surprised. I think the test should pass. I don't know why it fails, apparently it needs deeper investigation.

Reply:
@gromakovsky  For point 1, I have added the new error. 

For point 2, I discovered that we use wrong logic to check the operators in `transfer` entrypoint. I have added annotation and fix the bug as well as adding some regression tests. 

Continue from our discussion in slack: https://serokell.slack.com/archives/C019PCGKTK2/p1604402597008100?thread_ts=1604393896.004100&cid=C019PCGKTK2

The too big contract size error is still an issue even after removing proposalMetadata entrypoint. I have searched but no solution coming up. Maybe someone works at tezos could help. The final solution is probably to split this contract into 2 but I'm not sure how hard it will be.